### ISSUE TITLE
BUG: cluster: `fcluster` assigning all points to a single cluster with `"maxclust"` when `t>1`

### ISSUE THREAD

**Issue by gschivley** [scipy/scipy#12651]
I'm using `hac.linkage` and `hac.fcluster` to cluster points into varying levels (`t`) with `criterion=maxclust` and have found that with a small number of input features (`n`), individual points are not split into their own clusters as `t` approaches `n`. While the documentation does say that "no more than t flat clusters are formed", the behavior is still unintuitive. Is there something that I'm missing?

#### Reproducing code example:

```
import scipy.cluster.hierarchy as hac
import numpy as np

y = np.asarray([1, 4, 5]).reshape(-1, 1)
Z = hac.linkage(y)

hac.fcluster(Z=Z, t=3, criterion="maxclust")
```

The results are
```
array([1, 1, 1], dtype=int32)
```
I'd expect that when `t` equals `n`, the results would be the same as `hac.leaves_list(Z)`. Or that they should match the dendrogram.

![image](https://user-images.githubusercontent.com/10373332/89071478-d3a0f780-d344-11ea-86c9-b26cfab74b4c.png)

And here's a comparison with `AgglomerativeClustering` from sklearn - which I realize might be a different algorithm, but I would expect the same results for a simple example like this.

```
from sklearn.cluster import AgglomerativeClustering

y = np.asarray([1, 4, 5]).reshape(-1, 1)
AgglomerativeClustering(n_clusters=3, linkage="single").fit_predict(y)
```
With results:
```
array([1, 2, 0])
```

#### Scipy/Numpy/Python version information:

```
1.5.0 1.18.5 sys.version_info(major=3, minor=7, micro=6, releaselevel='final', serial=0)
```

**Comment by mdhaber** [comment]
@nmayorov I think you might be the most recent contributor to the _algorithms_ in this file (and yes, that was a while ago). I was wondering whether you'd be willing to look into this issue? I couldn't find references to the original algorithm, either in the original C code or the Cython translation (gh-3792). Any ideas for how to proceed?

**Comment by nmayorov** [comment]
As far as I remember I only worked on `linkage` function, its results is correct for this input.

The `fcluster(..., criterion="maxclust")` executes the following code:
```
    cdef double[:] max_dists = np.ndarray(n, dtype=np.float64)
    get_max_dist_for_each_cluster(Z, max_dists, n)
    # should use an O(n) algorithm
    cluster_maxclust_monocrit(Z, max_dists, T, n, mc)
```
The second argument of `cluster_maxclust_monocrit` must be a monotonic (increasing I suppose?) array (from the name and docstrings), but in this case `max_dists = [1, 3, 1]` (haven't figured out what it means exactly and if it's correct) is not monotonic.

So looks like there is some fundamental flaw or mismatch in this code path. I'll try to investigate more, but not sure what the outcome will be.

**Comment by nmayorov** [comment]
Another strange thing is that the public function `maxdists` allocates `n - 1` elements to compute these distances, whereas `fcluster` algorithm allocates `n` elements (code snippet above).

**Comment by nmayorov** [comment]
I've analyzed the code and found 2 bugs in `cluster_maxclust_monocrit`.

1. It does not consider the first threshold as potentially valid. It is a technical bug in the binary search implementation. It should be fixed as `lower_idx = 0 -> lower_idx = -1`. This is a known trick / approach in the binary search implementation, where one index is assigned as "invalid by convention" but the criterion function is never evaluated at it. This bug can be triggered if we set `t=2` for the data provided by the reporter.
2. When the number of clusters is >=n we should consider splits with even lower criterion, because splitting at the first distance will give at max `n - 1` clusters. This can be done by prepending an additional lower threshold in the array of thresholds or handling this case separately and assigning each original observation in its own cluster (easier approach to introduce).

I'll send the minimal fix + test for this bug report.
